### ISSUE TITLE
Could CrystalBall work on satisfiable or indeterminate  problems?

### ISSUE THREAD

**Issue by Geremia** [msoos/cryptominisat#676]
Could [CrystalBall](https://www.msoos.org/2019/06/crystalball-sat-solving-data-gathering-and-machine-learning/) work on satisfiable or indeterminate problems?

"[CrystalBall: SAT solving, Data Gathering, and Machine Learning](https://www.msoos.org/2019/06/crystalball-sat-solving-data-gathering-and-machine-learning/)" §"Enter DRAT":

>every unsatifiable problem that SAT solvers solve can be shown to be unsatisfiable through a set of operations called resolutions, that eventually lead to the equation 0=1, which is trivially false.

Couldn't DRAT also work for SAT or indeterminate problems, too?

Basically, I'm looking at how to analyze indeterminate problems to figure out what classes of clauses are causing "hang-ups".

**Comment by msoos** [comment]
`what classes of clauses are causing "hang-ups"` -- sorry to be a party popper, but there are no such classes of clauses. Let's say I create a CNF that encodes the AES encryption function:

AES(P, Key) = C

Now, this is trivially satisfiable, because the SAT solver will set some random plaintext P, set some random key K, and then execute the function, arriving at C. Trivial, solves in <1 second. Now, I add 128b of ciphertext (128 unit clauses), and 128b of plaintext (128 unit clauses), and run the solver. Now, the solver would need to break AES, finding the 128b key. If it could do that, I'd be a billionaire living on my own island. So it can't do that. But what are the classes of clauses that are causing the "hang-up"? That's a nonsensical question to ask. The whole CNF is causing it. Obviously not the  256 unit clauses we added. So the question is nonsensical.

Sorry, I can't help you with this, this sounds like you are trying to figure out cold fusion.

Closing.

Mate

**Comment by Geremia** [comment]
@msoos I'm referring to when you write that DRAT "can actually tell, after the solving has finished, which parts of the SAT solving were absolutely useless, and which ones useful." So by "parts of the SAT solving" you don't mean classes of CNF clauses.

**Comment by msoos** [comment]
I'm sorry I don't understand the question :( What I wrote is true, there are learnt clauses that are more useful than others (and so parts of the solving that are more useful than others)

But there are no classes of input clauses that are causing a "hang up" -- or more precisely, there are but they are what they are, e.g. the part that describes the AES. Of course you can add other clauses to the CNF and then only the AES part is problematic. But that's got nothing to do with learnt clauses, which is what Crystalball is about, right?


I'm sorry I seem to be not understanding your question. Please try to formulate the question differently, maybe it'll be clearer to me :)

**Comment by Geremia** [comment]
@msoos I think I need to learn more about DRAT, and I'll get back to you.

**Comment by msoos** [comment]
Cool! Do also check out FRAT, it's super cool:

https://github.com/digama0/frat

Also, sorry for my misunderstanding. I actually would REALLY welcome some work on Crystalball. I have even improved it in the meanwhile (unpublished currently). 


Further, I think FRAT is definitely nice and it could be used to make Crystalball even better I think!

Let me know if you really wanna work on this, we could have a short online call and maybe you can tell me what you wanna work on. Perhaps I can help.
